community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulator.java

```diff
@@ -0,0 +1,273 @@
+    }
+
+    @Override
+    public synchronized void close( boolean populationCompletedSuccessfully ) throws IOException
+    {
+        try
+        {
+            closeWriter();
+            if ( populationCompletedSuccessfully && failureBytes != null )
+            {
+                throw new IllegalStateException( "Can't mark index as online after it has been marked as failure" );
+            }
+            if ( populationCompletedSuccessfully )
+            {
+                assertPopulatorOpen();
+                markTreeAsOnline();
+            }
+            else
+            {
+                assertNotDropped();
+                ensureTreeInstantiated();
+                markTreeAsFailed();
+            }
+        }
+        finally
+        {
+            closeTree();
+        }
+    }
+
+    private void assertNotDropped()
+    {
+        if ( dropped )
+        {
+            throw new IllegalStateException( "Populator has already been dropped." );
+        }
+    }
+
+    @Override
+    public void markAsFailed( String failure ) throws IOException
+    {
+        failureBytes = failure.getBytes( StandardCharsets.UTF_8 );
+    }
+
+    private void ensureTreeInstantiated() throws IOException
+    {
+        if ( tree == null )
+        {
+            instantiateTree();
+        }
+    }
+
+    private void assertPopulatorOpen()
+    {
+        if ( tree == null )
+        {
+            throw new IllegalStateException( "Populator has already been closed." );
+        }
+    }
+
+    private void markTreeAsFailed() throws IOException
+    {
+        if ( failureBytes == null )
+        {
+            failureBytes = "".getBytes();
+        }
+        tree.checkpoint( IOLimiter.unlimited(), new FailureHeaderWriter( failureBytes ) );
+    }
+
+    private void markTreeAsOnline() throws IOException
+    {
+        tree.checkpoint( IOLimiter.unlimited(), ( pc ) -> pc.putByte( BYTE_ONLINE ) );
+    }
+
+    private <T extends Closeable> T closeIfPresent( T closeable ) throws IOException
+    {
+        if ( closeable != null )
+        {
+            closeable.close();
+        }
+        return null;
+    }
+
+    void closeWriter() throws IOException
+    {
+        singleWriter = closeIfPresent( singleWriter );
+    }
+
+    private void closeTree() throws IOException
+    {
+        tree = closeIfPresent( tree );
+    }
+
+    private class NativeSchemaIndexUpdater implements IndexUpdater
+    {
+        private boolean closed;
+
+        @Override
+        public void process( IndexEntryUpdate update ) throws IOException, IndexEntryConflictException
+        {
+            if ( closed )
+            {
+                throw new IllegalStateException( "Index updater has been closed." );
+            }
+            add( update );
+        }
+
+        @Override
+        public void remove( PrimitiveLongSet nodeIds ) throws IOException
+        {
+            throw new UnsupportedOperationException( "Implement me" );
```

Comment:
It will not be needed. That very method exists due to recovery issues in Lucene. The `GBPTree` recovers itself together with normal recovery w/o any additional logic.